source/common/websocket/codec.cc

```diff
@@ -0,0 +1,219 @@
+#include "source/common/websocket/codec.h"
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/scalar_to_byte_vector.h"
+
+namespace Envoy {
+namespace WebSocket {
+
+Encoder::Encoder() = default;
+
+void Encoder::newFrameHeader(const Frame& frame, std::vector<uint8_t>& output) {
+  // Set flags and opcode
+  pushScalarToByteVector(frame.flags_and_opcode_, output);
+
+  // Set payload length
+  if (frame.payload_length_ <= 125) {
+    // Set mask bit and 7-bit length
+    pushScalarToByteVector(frame.masking_key_ ? static_cast<uint8_t>(frame.payload_length_ | 0x80)
+                                              : static_cast<uint8_t>(frame.payload_length_),
+                           output);
+  } else if (frame.payload_length_ <= 65535) {
+    // Set mask bit and 16-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xfe : 0x7e), output);
+    // Set 16-bit length
+    for (uint8_t i = 1; i <= PAYLOAD_LENGTH_SIXTEEN_BIT; i++) {
```

Comment:
pushScalarToByteVector already knows how to add the individual bytes one at a time to the vector. Ca we do something like:

```
uint16_t len = frame.payload_length_;
pushScalarToByteVector(len, output)
```

Reply:
I tried it but it didn't work as expected when payload length is 256 test cases

Reply:
yes, endianness was the issue. `htobe16`, `htobe64` and `htobe32` could be used including other places.